R/FilterState-utils.R

```diff
@@ -78,213 +90,294 @@ init_filter_state <- function(x,
   if (is.null(varlabel)) varlabel <- character(0L)
 
   if (all(is.na(x))) {
-    return(
-      EmptyFilterState$new(
-        x = x,
-        x_reactive = x_reactive,
-        varname = varname,
-        varlabel = varlabel,
-        dataname = dataname,
-        extract_type = extract_type
-      )
+    args <- list(
+      x = x,
+      x_reactive = x_reactive,
+      dataname = dataname,
+      varname = varname,
+      choices = choices,
+      selected = selected,
+      varlabel = varlabel,
+      keep_na = keep_na,
+      keep_inf = keep_inf,
+      fixed = fixed,
+      extract_type = extract_type
     )
+    do.call(EmptyFilterState$new, args)
+  } else {
+    UseMethod("init_filter_state")
   }
-  UseMethod("init_filter_state")
 }
 
 #' @keywords internal
 #' @export
 init_filter_state.default <- function(x,
                                       x_reactive = reactive(NULL),
-                                      varname,
-                                      varlabel = attr(x, "label"),
-                                      dataname = NULL,
+                                      dataname,
+                                      varname = attr(x, "label"),
+                                      choices = unique(na.omit(x)),
+                                      selected = NULL,
+                                      keep_na = NULL,
+                                      keep_inf = NULL,
+                                      fixed = FALSE,
+                                      varlabel = character(0),
                                       extract_type = character(0)) {
   if (is.null(varlabel)) varlabel <- character(0)
-  InteractiveFilterState$new(
+
+  args <- list(
     x = x,
     x_reactive = x_reactive,
+    dataname = dataname,
     varname = varname,
+    choices = choices,
+    selected = selected,
     varlabel = varlabel,
-    dataname = dataname,
+    keep_na = keep_na,
+    keep_inf = keep_inf,
+    fixed = fixed,
     extract_type = extract_type
   )
+
+  do.call(InteractiveFilterState$new, args)
```

Comment:
there's lots of do.call's in the PR - is there a reason to use it here? If so I'd add a code comment

Reply:
The reason is two-fold: 1) to allow the `if`s to use the same argument list rather than typing out the full call twice, but to a different class at the beginning; 2) consistency.

Reply:
shame you can't just do:


init_filter_state.default <- function(x, ...) {
  InteractiveFilterState$new(x, ...)
}

and not type them out at all...

Reply:
I wasn't sure you couldn't but I have a feeling that would get us into trouble one way or another.

I'm willing to try in a different issue, though.

Reply:
There are mode `do.call` calls in the PR, those mostly deal with variable numbers of arguments.